nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -0,0 +1,765 @@
+                .map_err(ActionError::action_failed)?;
+        }
+    }
+
+    Ok(())
+}
+
+async fn ssc_send_snapshot_request_undo(
+    sagactx: ActionContext<SagaSnapshotCreate>,
+) -> Result<(), anyhow::Error> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let snapshot_id = sagactx.lookup::<Uuid>("snapshot_id")?;
+
+    let (.., disk) = LookupPath::new(&opctx, &osagactx.datastore())
+        .project_id(params.project_id)
+        .disk_name(&params.create_params.disk.clone().into())
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    // Delete any snapshots created by this saga
+    let datasets_and_regions = osagactx
+        .datastore()
+        .get_allocated_regions(disk.volume_id)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    for (dataset, region) in datasets_and_regions {
+        // Create a Crucible agent client
+        let url = format!("http://{}", dataset.address());
+        let client = CrucibleAgentClient::new(&url);
+
+        // Delete snapshot, it was created by this saga
+        warn!(log, "deleting snapshot {} {} {}", url, region.id(), snapshot_id);
+        client
+            .region_delete_snapshot(
+                &RegionId(region.id().to_string()),
+                &snapshot_id.to_string(),
+            )
+            .await
+            .map_err(|e| e.to_string())
+            .map_err(ActionError::action_failed)?;
+    }
+
+    Ok(())
+}
+
+async fn ssc_start_running_snapshot(
+    sagactx: ActionContext<SagaSnapshotCreate>,
+) -> Result<BTreeMap<String, String>, ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let snapshot_id = sagactx.lookup::<Uuid>("snapshot_id")?;
+
+    let (.., disk) = LookupPath::new(&opctx, &osagactx.datastore())
+        .project_id(params.project_id)
+        .disk_name(&params.create_params.disk.clone().into())
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    // For each dataset and region that makes up the disk, create a map from the
+    // region information to the new running snapshot information.
+    let datasets_and_regions = osagactx
+        .datastore()
+        .get_allocated_regions(disk.volume_id)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let mut map: BTreeMap<String, String> = BTreeMap::new();
+
+    for (dataset, region) in datasets_and_regions {
+        // Create a Crucible agent client
+        let url = format!("http://{}", dataset.address());
+        let client = CrucibleAgentClient::new(&url);
+
+        warn!(log, "dataset {:?} region {:?} url {}", dataset, region, url);
```

Comment:
`warn` doesn't seem like the right log level for these. Same elsewhere in this PR.

Reply:
fixed in d7e3c3f2